**Inclass: Python for Data Analysts**
- Part 1 of Data Analytics Specialization
- Course Length: 12 hours
- Instructor : Fafilia Masrofin
- Last Updated: September 2022

This notebook was made based on main materials `1_Python_for_Data_Analysts.ipynb`
___

In [37]:
1+1

2

Markdown Cell = text

# Test Capital Letter

# Top-Down Approach 

The coursebook is part of the **Data Analytics Specialization** offered by [Algoritma](https://algorit.ma). It takes a more accessible approach compared to Algoritma's core educational products, by getting participants to overcome the "how" barrier first, rather than a detailed breakdown of the "why". 

This translates to an overall easier learning curve, one where the reader is prompted to write short snippets of code in frequent intervals, before being offered an explanation on the underlying theoretical frameworks. Instead of mastering the syntactic design of the Python programming language, then moving into data structures, and then the `pandas` library, and then the mathematical details in an imputation algorithm, and its code implementation; we would do the opposite: Implement the imputation, then a succinct explanation of why it works and applicational considerations (what to look out for, what are assumptions it made, when _not_ to use it etc).

For the most part, experience in Python programming is good to have but not required. Familiarity with data manipulation and data structures in a different programming language a welcome addition but again, not required.

## Training Objectives

This coursebook is intended for participants new to the world of data analysis and / or programming. No prior programming knowledge is assumed. 

The coursebook focuses on:
- Introduction to the `pandas` library. 
- Introduction to `DataFrame`  
- Data Types
- Exploratory Data Analysis I
- Indexing and Subsetting

The final part of this course is a Graded Asssignment, where you are expected to apply all that you've learned on a new dataset, and attempt the given questions.

# Environment

**Cara membuat virtual environment baru:**

1. Membuka Anaconda Prompt

2. Buat environment dengan nama `ENV_NAME`:
    ```
    conda create -n <ENV_NAME> python=<PYTHON_VERSION>
    ```
    Contoh: `conda create -n theia_da python=3.9`


3. Mengaktifkan (pindah) environment:
    ```
    conda activate <ENV_NAME>
    ```
    Contoh: `conda activate theia_da`
    

4. Memasang kernel di dalam environment:
    ```
    pip install ipykernel
    python -m ipykernel install --user --name=<ENV_NAME>
    ```
    Contoh: `python -m ipykernel install --user --name=theia_da`

5. Cara instalasi package:
    ```
    pip install <PACKAGE_NAME>
    ```
    Contoh: `pip install pandas`
    
**Perintah `conda` yang sering digunakan:**

- `conda env list`: cek daftar **environment** pada Anaconda
- `conda list`: cek daftar **package** di dalam sebuah environment

Referensi perintah lain: https://docs.conda.io/projects/conda/en/latest/commands.html

**❓ Knowledge Check:**

Pilihlah jawaban yang tepat dengan memberikan tanda centang pada kotak. Jawaban bisa lebih dari satu.

1. Berikut yang merupakan Package Manager untuk instalasi package adalah ...

    - [ ] kernel
    - [x] conda
    - [x] pip
    - [ ] environment

2. Kita perlu menggunakan ...... agar sebuah virtual environment dapat digunakan pada Jupyter Notebook.

    - [x] kernel
    - [ ] package manager
    - [ ] library / package
    - [ ] IDE

3. Berikut adalah tujuan kita membuat virtual environment, **KECUALI** ...

    - [ ] Kolaborasi (sharing environment)
    - [ ] Deliver/Deploy aplikasi ke user
    - [ ] Mengisolasi package beserta versi-nya
    - [x] Virtual environment harus dibuat agar Python dapat dijalankan

# Introduction to Jupyter Notebook

## Markdown Cell and Code Cell

Kumpulan shortcut: **CTRL + SHIFT + P**

Tipe cell dalam notebook:
1. Markdown
2. Code
    
Di bawah ini adalah `cell`. Kita dapat memberikan format pada teks seperti **bold** atau *italic*.

In [38]:
print('and this is code cell where you put your python codes down')

and this is code cell where you put your python codes down


## Command Mode and Edit Mode

Mode cell dalam notebook:
1. Command mode (cell berwarna BIRU)
    - B: menambahkan cell baru di Bawah (Below)
    - A: menambahkan cell baru di Atas (Above)
    - DD: Delete cell
    - C: Copy cell
    - V: Paste cell
    - Y: Mengubah ke code cell
    - M: Mengubah ke markdown cell
    - Enter: Mengubah command mode menjadi edit mode


2. Edit mode (cell berwarna HIJAU)
    - Ctrl + Enter: eksekusi satu cell
    - Esc: Mengubah edit mode menjadi command mode

# Introduction to Python

## Variables and Keywords

**Variable** adalah sebuah nama yang dipakai untuk menunjukkan sebuah nilai. Tanda `=` dipakai untuk membuat variable baru. Proses ini sering disebut sebagai **assignment**.

In [39]:
angka = 5

In [40]:
angka

5

In [41]:
# print(angka)


Python adalah bahasa pemrograman yang **case-sensitive** sehingga penamaan variable menjadi hal yang perlu diperhatikan. Misal penulisan variable `activity` dengan awalan huruf `a` kecil berbeda dengan yang diawali dengan huruf `A` besar.

In [42]:
activity = 'programming'

In [43]:
print(activity)

programming


In [44]:
# print activity



In [45]:
# activity == Activity -> kode ini akan menghasilkan error karena activity dengan letter case "a" berbeda dengan Activity
# dengan upper case "A"

In [46]:
activity == 'programming'

True

Berikut beberapa anjuran dalam memberikan nama variable pada Python:
- Menggunakan kombinasi dari huruf kapital (A-Z), huruf nomina (a-z), angka (0-9).
- Special character `!, $ , &, dll` tidak dapat digunakan dalam penamaan variabel.
- Tidak boleh menggunakan angka di awal.
- Tidak boleh menggunakan keyword pada Python
- Bersifat case-sensitive sehingga penamaan variable `algoritma`, `ALGORITMA`, dan `Algoritma` adalah 3 variable yang berbeda

**Keywords** adalah kata kunci yang sudah ditetapkan oleh Python sebagai nama yang tidak bisa dipakai baik untuk penamaan fungsi, variabel, dan lainnya. Keyword ditulis dalam lower-case (huruf kecil semua) kecuali keyword `True`, `False`, dan `None`. Sejauh ini (Python 3.8) keyword yang ada pada Python adalah sebagai berikut:

In [47]:
#Cek daftar keyword
import keyword
keyword.kwlist

['False',
 'None',
 'True',
 '__peg_parser__',
 'and',
 'as',
 'assert',
 'async',
 'await',
 'break',
 'class',
 'continue',
 'def',
 'del',
 'elif',
 'else',
 'except',
 'finally',
 'for',
 'from',
 'global',
 'if',
 'import',
 'in',
 'is',
 'lambda',
 'nonlocal',
 'not',
 'or',
 'pass',
 'raise',
 'return',
 'try',
 'while',
 'with',
 'yield']

Untuk membuktikan bahwa keyword tidak dapat digunakan sebagai nama variabel, mari kita coba untuk menyimpan nilai 1 pada variabel `True`.

In [48]:
if angka < 5 : print("ok") 
else : print ("no")

no


## Python Data Types

### String

Python mewakili text apa pun sebagai objek `str`. Contohnya kata "hamburger" dan frasa "saya suka hamburger" sama-sama merupakan sebuah string. Ada beberapa cara untuk membuat nilai string:

- menggunakan `'text'` (yaitu: `'Hello World!'`)
- menggunakan `"text"` (yaitu : `"I'm a programmer"`)
- menggunakan `'''text'''` atau `"""text"""` (yaitu: `''' "I'm smart!", he said '''`)

In [49]:
# code here


### Number

Untuk menyimpan data bersifat numerik, python memiliki dua tipe data asli yang disebut `int` dan `float`.
- `int` digunakan untuk menyimpan bilangan bulat (yaitu: 1,2,-3)
- `float` digunakan untuk menyimpan bilangan real (yaitu: 0.7, -1.8, -1000.0).

In [50]:
a = 123445678910
b = 0.6724624632746

In [51]:
type(a)

int

In [52]:
type(b)

float

In [53]:
print(a)
print(b)

123445678910
0.6724624632746


**Operasi Angka** \
Operator Aritmatika:
- `+` - Penambahan
- `-` - Pengurangan
- `*` - Perkalian
- `/` - Divisi
- `//` - Pembagian dengan Pembulatan
- `%` - Modul
- `**` - Eksponen

Arithmetical Operators

In [54]:
1 + 1

2

In [55]:
5 / 2

2.5

In [56]:
# code here


Operasi perbandingan:
- `==` : Untuk melihat apakah object di kiri dan kanan sama
- `!=` : Untuk melihat apakah object di kiri dan kanan **tidak** sama
- `>` : Untuk melihat apakah object di kiri lebih besar dari object yang di kanan
- `>=` : Untuk melihat apakah object di kiri lebih besar atau **sama dengan** dari object yang di kanan
- `<` : Untuk melihat apakah object di kiri lebih kecil dari object yang di kanan
- `<=` : Untuk melihat apakah object di kiri lebih kecil atau **sama dengan** dari object yang di kanan

Comparison Operators

In [57]:
satu = 1
dua = 2

In [58]:
satu <= dua

True

In [59]:
satu == dua

False

In [60]:
satu != dua

True

In [61]:
# code here


**🧠 Knowledge Check :**

If your friend give you a budget of USD 97.5, and ask you to buy candies that cost USD 0.75 each and distribute them **fairly** to 17 children, what is the maximum number of candy each child can get ? 

Try to use your python knowledge to solve this problem

In [62]:
97.5 (0.75 * 17)


<>:1: SyntaxWarning: 'float' object is not callable; perhaps you missed a comma?
<>:1: SyntaxWarning: 'float' object is not callable; perhaps you missed a comma?
C:\Users\muham\AppData\Local\Temp\ipykernel_11088\3287648229.py:1: SyntaxWarning: 'float' object is not callable; perhaps you missed a comma?
  97.5 (0.75 * 17)
C:\Users\muham\AppData\Local\Temp\ipykernel_11088\3287648229.py:1: SyntaxWarning: 'float' object is not callable; perhaps you missed a comma?
  97.5 (0.75 * 17)
C:\Users\muham\AppData\Local\Temp\ipykernel_11088\3287648229.py:1: SyntaxWarning: 'float' object is not callable; perhaps you missed a comma?
  97.5 (0.75 * 17)


TypeError: 'float' object is not callable

### Boolean

Boolean menyimpan nilai yang sangat sederhana dalam komputer dan pemrograman, `True` atau `False`.

In [ ]:
x = True 
y = False 

In [ ]:
x

### List

`list` digunakan untuk menyimpan beberapa nilai dalam python. Untuk membuatnya, cukup letakkan nilai di dalam tanda kurung (yaitu: `x = [1,2,3]` )

In [ ]:
number = [1,2,3,4,5]
names = ["andi","budi","caca","deni","eca"]

In [ ]:
print(number)
print(names)

In [ ]:
number.append(6) 

In [ ]:
number

**Operasi List**
- `x.append(a)` : tambahkan a ke x
- `x.remove(a)` : hapus a dari x

Selain operator yang dikenal sebelumnya, salah satu list yang paling berguna adalah dengan menerapkan fungsi agregasi seperti:
- `len(x)` : ekstrak panjang daftar
- `a in b` : memeriksa apakah nilai `a` ada di objek daftar `b`
- `max(x)` : mendapatkan nilai tertinggi dalam x
- `sum(x)` : mendapatkan jumlah nilai dalam x

Operasi lain yang harus diketahui dalam daftar adalah pengindeksan:
- `x[i]` : mengakses elemen ke-i dari x

**🧠Knowledge Check** : 

What is the average value of `number` ? 

In [ ]:
# code here


# Introduction to Pandas Library

## Working with DataFrame

`pandas` adalah library yang powerful sebagai tools analisis data dan struktur pada Python. Dengan `pandas`, mengolah data menjadi mudah karena disediakan salah satu objek bernama **DataFrame**. Dengan dataframe kita dapat membaca sebuah file, mengolah suatu data dengan menggunakan operasi seperti join, distinct, group by, agregasi, dan teknik lainnya.

> Lebih lengkapnya silahkan kunjungi [official documentation](https://pandas.pydata.org/)

Untuk menggunakan `pandas`, kita perlu import terlebih dahulu library dengan cara berikut ini:

In [ ]:
import pandas as pd


Kita bisa menggunakan teknik **aliasing** agar pengetikan nama library tidak terlalu panjang, yaitu dengan `as`.

In [ ]:
##code here
pd.__version__


Semua method pada `pandas` dapat dipanggil dengan syntax seperti: `pandas.function_name()`. Langkah pertama yang akan kita lakukan adalah membaca data. Kita dapat menggunakan method `.read_csv()` untuk membaca sebuah file dengan format `.csv`.

In [ ]:
##code here
pd.read_csv("data_input/rice.csv")

Kita coba set kolom pertama `Unnamed: 0` menjadi index dari dataframe, dengan cara menambahkan **parameter** `index_col=0`.

In [ ]:
##code here
pd.read_csv("data_input/rice.csv",index_col=0)


In [ ]:
rice['receipt_id']

Intuisi dari parameter `index_col` pada `read_csv()` adalah menjadikan kolom pada Dataframe sebagai index pada baris. Berikut beberapa nilai yang dapat ditampung oleh parameter `index_col`.
- (0, 1, 2, dst) : Menunjukkan index kolom yang akan dijadikan sebagai index baris.
- `'nama_kolom'` : Selain menggunakan nilai index nya, kita juga dapat langsung mengetikkan nama kolomnya.
- `False` : Menggunakan index yang disediakan oleh `pandas`.

**Additional Information:**

1. Python menggunakan sistem **zero based indexing** yang berarti, index pada python dimulai dari angka 0.
2. Terdapat beberapa parameter dalam method `.read_csv()`, dokumentasi selengkapnya ada [di sini](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html)

🧠 **Knowledge Check: Error**

Mengacu pada konsep penulisan **python keywords** diatas, manakah diantara ke-4 kode program ini yang **tidak** menghasilkan error ketika dijalankan?

- [x] `pd.read_csv("data_input/rice.csv", index_col=false)`
- [x] `Import pandas as pd`
- [ ] `print(100-2)`
- [x] `None = 2`

In [ ]:
#your code here



**What you need to know**:
- Python, sama dengan bahasa pemrograman lainnya adalah bersifat case sensitive. `Sales` dan `sales` akan diidentifikasikan sebagai object yang berbeda.
- Python keywoard tidak dapat digunakan sebagai nama variable
- Saat memberikan penamaan pada variable, awali dengan huruf dan gunakan underscore (`_`) untuk menghubungkan antar kata.
    - Salah: `2022`, `2022sales`, `sales-2022`, `sales.2022`
    - Benar: `sales_2019`, `profit_after_tax`

#### 🧠 Dive Deeper

1. Baca kembali data `rice` menggunakan perintah `pd.read_csv()`, kemudian simpan kedalam variable baru bernama `rice_new`. Dengan menggunakan parameter `index_col`, jadikan kolom `receipt_id` sebagai index baris nya!
2. `pandas.DataFrame.head(n)` dapat digunakan untuk menampilkan sebagian data teratas, dengan asumsi bahwa nilai `n` adalah jumlah baris yang ingin kita tampilkan. Silahkan set `head()` dengan nilai `n=8` pada data `rice_new`, kemudian lihat apa yang terjadi! 
3. Lawan dari `head()` adalah `tail()`. Method `tail()` akan menampilkan data dari urutan paling bawah. Silahkan set `tail()` dengan nilai `n=4`, kemudian lihat apa yang terjadi!

In [ ]:
#your code here
rice_new = pd.read_csv("data_input/rice.csv", index_col=1) 

In [77]:
rice = pd.read_csv("data_input/rice.csv", index_col=1) 

In [ ]:
rice_new

In [ ]:
#your code here
rice_new


In [ ]:
rice_new.head(6)

In [ ]:
rice_new.tail(8)

## Tipe Data `pandas`

Dataframe terdiri dari beberapa **Series** (mengacu pada satu kolom). Dalam satu series harus memiliki satu tipe data yang sama. `pandas`akan mencoba untuk infer tipe data dari masing-masing Series, tapi tidak selalu benar.

Cara cek tipe data: `dtypes` atau `.info()` untuk lebih lengkapnya

**_Soal 1:_** Dengan menggunakan data rice, silahkan cek tipe data menggunakan `dtypes`!

In [ ]:
##code here
rice_new.info()


**Notes:**

- `int64`: integer (bilangan bulat, tanpa koma)
- `float64`: bilangan desimal (berkoma)
- `object`: text (string)

🧠 **Knowledge check: `.dtypes` and pandas attributes**

Salin kode program di bawah ini, kemudian jalankan. 
```
x = [2019, 4, 'data science']
x.dtypes

```
Apa yang terjadi?

In [ ]:
##code here
x = [2019, 4, 'data science']


Ketikkan perintah `type(x)`, maka apa perbedaannya dengan sebelumnya?

In [ ]:
##code here

type(x)

Kesimpulan: Atribut `dtypes` hanya dimiliki oleh objek dataframe sehingga `dtypes` tidak bisa diaplikasikan untuk list.

Perhatikan dataframe baru di bawah ini untuk contoh lain dari tipe data pada `pandas`!

In [ ]:
employees = pd.DataFrame({
    'name': ['Anita', 'Brian'],
    'age': [34, 29],
    'joined': [pd.Timestamp('20190410'), pd.Timestamp('20171128')],
    'degree': [True, False],
    'hourlyrate': [35.5, 29],
    'division': ['HR', 'Product']
})
employees.dtypes

In [ ]:
employees
employees.dtypes

- `name` [`object`]: store text values
- `age` [`int`]: integer values
- `joined` [`datetime`]: date and time values
- `degree` [`bool`]: True/False values
- `hourlyrate` [`float`]: floating point values
- `division` [`object`]: store text values

### Categorical and Numerical Variables

Tipe data categorical adalah tipe data yang memiliki karakteristik dimana nilainya dapat berulang pada sebuah Series di dataframe. 

Dua alasan mengapa kita perlu menggunakan tipe data categorical:

1. Dari sisi "business perspective", hal ini dapat menginformasikan dan memandu seorang Analyst pada pertanyaan seperti metode statistik atau tipe plot mana yang digunakan untuk mengolah data.

2. Dari sisi teknikal, ketika kita bekerja dengan tipe data categorical pada pandas, hal ini akan jauh menghemat memori dan menambah kecepatan komputasional.

Dikutip dari official documentation:

> Categoricals are a pandas data type corresponding to categorical variables in statistics. A categorical variable takes on a limited, and usually fixed, number of possible values (categories; levels in R). Examples are gender, social class, blood type, country affiliation or rating via Likert scales.

Untuk mengubah tipe data ke categorical pada pandas, Anda dapat melakukannya dengan method `astype()` berikut:

**Formula** 
```
df['column_name'] = df['column_name'].astype('new_data_types')
```

**Example**
```
# convert marital_status to category
employees['marital_status'] = employees['marital_status'].astype('category')

# convert experience to integer
employees['experience'] = employees['experience'].astype('int')
```


**_Soal 2:_** Perhatikan dataframe `employees`. Adakah kolom yang tipe datanya belum tersimpan dengan benar? Ubahlah ke tipe data yang tepat dengan menggunakan method `astype()`!

In [ ]:
employees['division']=employees['division'].astype('category')
employees.dtypes

In [ ]:
employees = employees.drop("divison", axis='columns')

**Notes**

- `type()` adalah sebuah **fungsi** pada python. Fungsi `type()` digunakan untuk mengecek struktur/tipe objek atau variabel dimana penggunaannya tidak didahului dengan pemanggilan nama objek.
- `dtypes` adalah **atribut** yang dimiliki oleh objek dataframe. `dtypes` digunakan untuk melihat tipe data dari masing-masing kolom yang ada pada dataframe. Penggunaan `dtypes` diawali dengan memanggil nama objeknya (karena dia adalah atribut sebuah objek)

In [ ]:
rice['receipt_id'].astype('int64')

### Rangkuman Data Types

Fokus pada kolom Pandas dtype dan usage-nya:

| Pandas dtype  | Python type  | NumPy type                                                     | Usage                                        |
|---------------|--------------|----------------------------------------------------------------|----------------------------------------------|
| object        | str or mixed | string_, unicode_, mixed types                                 | Text or mixed numeric and non-numeric values |
| int64         | int          | int_, int8, int16, int32, int64, uint8, uint16, uint32, uint64 | Integer numbers                              |
| float64       | float        | float_, float16, float32, float64                              | Floating point numbers                       |
| bool          | bool         | bool_                                                          | True/False values                            |
| datetime64    | NA           | datetime64[ns]                                                 | Date and time values                         |
| timedelta[ns] | NA           | NA                                                             | Differences between two datetimes            |
| category      | NA           | NA                                                             | Finite list of text values                   |

Referensi: [Overview of Pandas Data Types](https://pbpython.com/pandas_dtypes.html)

## Exploratory Data Analysis Tools :

Exploratory Data Analysis (**EDA**) mengacu pada proses melakukan investigasi awal pada data, seringkali dengan tujuan untuk lebih mengenal dengan karakteristik data tertentu. EDA dilakukan dengan bantuan ringkasan statistik dan teknik grafis sederhana untuk melihat struktur data yang kita miliki.

Beberapa tools sederhana pada `pandas` yang dapat digunakan untuk melakukan EDA adalah sebagai berikut:
- `.head()` and `.tail()`
- `.describe()`
- `.shape` and `.size`
- `.axes`
- `.dtypes`

### `head()` and `tail()`

- `head(n)` digunakan untuk inspeksi `n` data teratas dari sebuah dataframe.
- `tail(n)` digunakan untuk inspeksi `n` data terbawah dari sebuah dataframe.

In [ ]:
##code here



In [ ]:
##code here



### `describe()`

Method `describe()` menampilkan 8 ringkasan statistika deskriptif. Secara default menampilkan ringkasan untuk kolom numerik. 

Ringkasan statistika yang dimaksud adalah sebagai berikut:
- Count: banyaknya baris pada dataframe
- Mean: rata-rata nilai
- Standard Deviation: jarak rata-rata antara data ke mean (titik pusat data)
- Minimum Value: nilai terkecil dari keseluruhan data
- 25th Percentile (Q1)
- 50th Percentile (Q2/Median)
- 75th Percentile (Q3)
- Maximum Value: nilai terbesar dari keseluruhan data

In [ ]:
##code here



Ringkasan seperti count, mean, std, min, dan max memiliki method nya masing-masing:

In [ ]:
##code here



Untuk ringkasan seperti Q1, Q2, Q3 kita dapat menggunakan method `quantile()`. Secara default adalah nilai Q2.

In [ ]:
##code here



**Additional Information**

Kita bisa menambahkan parameter `include` ataupun `exclude` pada `describe()` untuk melihat statistika deskriptif dari variable non-numeric:

In [67]:
##code here
rice.describe(include='object')


purchase_time category sub_category      format yearmonth
count            12000    12000        12000       12000     12000
unique           11609        1            1           3        12
top     6/14/2018 0:00     Rice         Rice  minimarket   2018-07
freq                 4    12000        12000        7088      1000

- unique: nilai unik di setiap kolom atau variabel
- top: nilai terbanyak (modus)
- freq: banyaknya kemunculan modus

Misal kita tertarik melihat unique pada kolom `format`:

In [ ]:
##code here



**_Soal 3:_** Melihat statistika deskriptif untuk kolom bertipe data **selain object**  menggunakan `exclude`

In [63]:
##code here

rice.describe(exclude='object')

Unnamed: 0    receipt_id  receipts_item_id     unit_price  \
count  12000.00000  1.200000e+04      1.200000e+04   12000.000000   
mean    6000.50000  7.650135e+06      2.457950e+07   70013.146313   
std     3464.24595  1.873838e+06      7.171105e+06   29905.391437   
min        1.00000  3.173994e+06      9.282023e+06    9395.000000   
25%     3000.75000  5.983209e+06      1.820694e+07   61900.000000   
50%     6000.50000  7.443618e+06      2.234337e+07   63500.000000   
75%     9000.25000  9.149786e+06      3.108379e+07   66000.000000   
max    12000.00000  1.146321e+07      3.842939e+07  219400.000000   

            discount      quantity  
count   12000.000000  12000.000000  
mean      835.305750      1.332917  
std      6207.475704      0.980304  
min     -4100.000000      1.000000  
25%         0.000000      1.000000  
50%         0.000000      1.000000  
75%         0.000000      1.000000  
max    320000.000000     19.000000

**_Soal 4:_** Melihat statistika deskriptif untuk kolom bertipe data **numeric** menggunakan `include`

In [64]:
##code here

rice.describe(include='number')

Unnamed: 0    receipt_id  receipts_item_id     unit_price  \
count  12000.00000  1.200000e+04      1.200000e+04   12000.000000   
mean    6000.50000  7.650135e+06      2.457950e+07   70013.146313   
std     3464.24595  1.873838e+06      7.171105e+06   29905.391437   
min        1.00000  3.173994e+06      9.282023e+06    9395.000000   
25%     3000.75000  5.983209e+06      1.820694e+07   61900.000000   
50%     6000.50000  7.443618e+06      2.234337e+07   63500.000000   
75%     9000.25000  9.149786e+06      3.108379e+07   66000.000000   
max    12000.00000  1.146321e+07      3.842939e+07  219400.000000   

            discount      quantity  
count   12000.000000  12000.000000  
mean      835.305750      1.332917  
std      6207.475704      0.980304  
min     -4100.000000      1.000000  
25%         0.000000      1.000000  
50%         0.000000      1.000000  
75%         0.000000      1.000000  
max    320000.000000     19.000000

**_Soal 5:_** Bagaimana menampilkan statistika deskriptif hanya kolom yang bertipe data **int64** dan **object** saja?

In [66]:
##code here
rice.describe(include=['int64','float64'])


Unnamed: 0    receipt_id  receipts_item_id     unit_price  \
count  12000.00000  1.200000e+04      1.200000e+04   12000.000000   
mean    6000.50000  7.650135e+06      2.457950e+07   70013.146313   
std     3464.24595  1.873838e+06      7.171105e+06   29905.391437   
min        1.00000  3.173994e+06      9.282023e+06    9395.000000   
25%     3000.75000  5.983209e+06      1.820694e+07   61900.000000   
50%     6000.50000  7.443618e+06      2.234337e+07   63500.000000   
75%     9000.25000  9.149786e+06      3.108379e+07   66000.000000   
max    12000.00000  1.146321e+07      3.842939e+07  219400.000000   

            discount      quantity  
count   12000.000000  12000.000000  
mean      835.305750      1.332917  
std      6207.475704      0.980304  
min     -4100.000000      1.000000  
25%         0.000000      1.000000  
50%         0.000000      1.000000  
75%         0.000000      1.000000  
max    320000.000000     19.000000

### `shape` dan `size`

`shape` dan `size` adalah atribut dari sebuah dataframe yang memberikan informasi terkait dimensi data dan ukuran data

In [68]:
##code here
rice.shape


(12000, 11)

In [69]:
rice.shape[0]

12000

In [70]:
rice.shape[1]

11

Angka pertama menunjukkan banyaknya row/observasi, angka kedua menunjukkan banyaknya kolom.

In [71]:
##code here

rice.size

132000

`size` = banyaknya row dikali banyaknya kolom.

### `axes`

`axes` adalah atribut dataframe yang memberikan informasi terkait index dataframe (baik index kolom maupun index baris)

In [72]:
##code here
rice.axes


[RangeIndex(start=0, stop=12000, step=1),
 Index(['Unnamed: 0', 'receipt_id', 'receipts_item_id', 'purchase_time',
        'category', 'sub_category', 'format', 'unit_price', 'discount',
        'quantity', 'yearmonth'],
       dtype='object')]

In [75]:
rice.axes[1]

Index(['Unnamed: 0', 'receipt_id', 'receipts_item_id', 'purchase_time',
       'category', 'sub_category', 'format', 'unit_price', 'discount',
       'quantity', 'yearmonth'],
      dtype='object')

- `axes[0]`: melihat index baris
- `axes[1]`: melihat index kolom

In [ ]:
##code here



### `dtypes`

Atribut `dtypes` digunakan untuk melakukan inspeksi tipe data.

In [79]:
rice.head()

Unnamed: 0  receipts_item_id    purchase_time category  \
receipt_id                                                           
9622257              1          32369294  7/22/2018 21:19     Rice   
9446359              2          31885876  7/15/2018 16:17     Rice   
9470290              3          31930241  7/15/2018 12:12     Rice   
9643416              4          32418582   7/24/2018 8:27     Rice   
9692093              5          32561236  7/26/2018 11:28     Rice   

           sub_category       format  unit_price  discount  quantity yearmonth  
receipt_id                                                                      
9622257            Rice  supermarket    128000.0         0         1   2018-07  
9446359            Rice   minimarket    102750.0         0         1   2018-07  
9470290            Rice  supermarket     64000.0         0         3   2018-07  
9643416            Rice   minimarket     65000.0         0         1   2018-07  
9692093            Rice  supermarket    124500.0         0         1   2018-07

In [101]:
##code here

rice.dtypes

Unnamed: 0             int64
receipts_item_id       int64
purchase_time         object
category            category
sub_category        category
format              category
unit_price           float64
discount               int64
quantity               int64
yearmonth             object
dtype: object

**_Soal 6:_** Kolom mana yang tipe datanya belum sesuai? Silahkan ubah dengan method `.astype()`!

In [97]:
##code here
# rice.purchase_time = rice.purchastime.astype()

In [98]:
rice['sub_category'] = rice['sub_category'].astype('category')
rice['format'] = rice['format'].astype('category')
rice['category'] = rice['category'].astype('category')

In [112]:
rice[['sub_category', 'format', 'category']] = rice[['sub_category', 'format', 'category']].astype('category')

In [113]:
rice['purchase_time'] = rice['purchase_time'].astype('datetime64')

In [115]:
rice['yearmonth'] = rice['yearmonth'].astype('datetime64')

In [96]:
rice.dtypes

Unnamed: 0             int64
receipts_item_id       int64
purchase_time         object
category            category
sub_category        category
format              category
unit_price           float64
discount               int64
quantity               int64
yearmonth             object
dtype: object

In [91]:
rice.tail()

Unnamed: 0  receipts_item_id     purchase_time category  \
receipt_id                                                            
5760491          11996          17555486  12/15/2017 21:06     Rice   
5598782          11997          16999147   12/2/2017 14:12     Rice   
5735850          11998          17434503  12/13/2017 19:17     Rice   
5678748          11999          17317935   12/8/2017 22:04     Rice   
5702411          12000          17341559  12/10/2017 17:04     Rice   

           sub_category       format  unit_price  discount  quantity yearmonth  
receipt_id                                                                      
5760491            Rice  supermarket    128000.0         0         1   2017-12  
5598782            Rice   minimarket     64000.0         0         1   2017-12  
5735850            Rice  supermarket     59990.0      3000         1   2017-12  
5678748            Rice   minimarket     59500.0      3000         1   2017-12  
5702411            Rice   minimarket     59500.0      3000         1   2017-12

🧠 **Knowledge Check: Data types**

Misalkan saja kita memiliki sebuah DataFrame dengan nama `inventory`.

1. Ketika menjalankan perintah `inventory.dtypes`, maka pandas akan menampilkan tipe data pada setiap kolom (series). Manakah kolom dibawah ini yang mungkin memiliki tipe data yang salah?
 - [ ] `units_instock`: int64
 - [ ] `discount_price`: float64
 - [ ] `item_name`: object
 - [ ] `units_sold`: object
 
 
2. Kita ingin mengetahui jumlah kolom yang terdapat pada `inventory` DataFrame, manakah dibawah ini perintah yang tepat untuk menampilkan jumlah kolom `inventory`? Pilih beberapa kode program yang mungkin!
 - [x] `print(len(inventory.columns))`
 - [x] `print(inventory.shape[1])`
 - [ ] `print(len(rice.axes[1]))`

In [102]:
inventory = pd.DataFrame({
    'units_instock': [50, 40, 30],
    'discount_price': [15, 5, 7],
    'item_name': ['bawang', 'garam', 'gula'],
    'unit_sold': ['123', '456', '789']
})
inventory.dtypes

units_instock      int64
discount_price     int64
item_name         object
unit_sold         object
dtype: object

In [110]:
##code here
inventory['unit_sold'] = inventory['unit_sold'].astype('int64')


In [ ]:
##code here



---
## Indexing and Subsetting with Pandas

Indexing digunakan untuk memilih dan mengambil sebagian data yang hanya diperlukan dalam proses analisa data yang sedang dikerjakan. Contohnya:
- Compare sales pada tahun 2018 vs 2019
- Identifikasi peluang penjualan pada segment pasar (ex : Wholesale vs Retail)
- Melihat quarter terbaik untuk setiap tahun yang dapat digunakan untuk tujuan promosi
- dan sebagainya

Untuk melakukan indexing dan subsetting (slicing data) pada `pandas`, kita dapat menggunakan cara sebagai berikut:
- `head()` and `tail()`  
- `select_dtypes()`  
- Using `.drop()` 
- The `[]` operator
- `.loc`  
- `.iloc`
- Conditional subsetting

### `select_dtypes()`

Method `select_dtypes()` digunakan untuk memilih kolom sesuai dengan tipe datanya. Ada 2 parameter yang dapat digunakan di dalam method `select_dtypes()` yaitu parameter `include` dan `exclude` (seperti pada `describe()`.

Misal:
- parameter `include = 'category'` artinya kita memilih semua kolom dengan tipe data 'category'
- sebaliknya, ketika menggunakan parameter `exclude = 'category'` maka kolom-kolom dengan tipe data selain 'category' akan ditampilkan.


In [118]:
#your code here
rice.select_dtypes(include='category').head()


category sub_category       format
receipt_id                                   
9622257        Rice         Rice  supermarket
9446359        Rice         Rice   minimarket
9470290        Rice         Rice  supermarket
9643416        Rice         Rice   minimarket
9692093        Rice         Rice  supermarket

### `drop()`

Method `drop()` digunakan untuk membuang baris atau kolom yang tidak ingin digunakan untuk tujuan analisis. Secara default method ini akan menghapus baris

In [119]:
#your code here
rice.drop(columns = 'format')


Unnamed: 0  receipts_item_id       purchase_time category  \
receipt_id                                                              
9622257              1          32369294 2018-07-22 21:19:00     Rice   
9446359              2          31885876 2018-07-15 16:17:00     Rice   
9470290              3          31930241 2018-07-15 12:12:00     Rice   
9643416              4          32418582 2018-07-24 08:27:00     Rice   
9692093              5          32561236 2018-07-26 11:28:00     Rice   
...                ...               ...                 ...      ...   
5760491          11996          17555486 2017-12-15 21:06:00     Rice   
5598782          11997          16999147 2017-12-02 14:12:00     Rice   
5735850          11998          17434503 2017-12-13 19:17:00     Rice   
5678748          11999          17317935 2017-12-08 22:04:00     Rice   
5702411          12000          17341559 2017-12-10 17:04:00     Rice   

           sub_category  unit_price  discount  quantity  yearmonth  
receipt_id                                                          
9622257            Rice    128000.0         0         1 2018-07-01  
9446359            Rice    102750.0         0         1 2018-07-01  
9470290            Rice     64000.0         0         3 2018-07-01  
9643416            Rice     65000.0         0         1 2018-07-01  
9692093            Rice    124500.0         0         1 2018-07-01  
...                 ...         ...       ...       ...        ...  
5760491            Rice    128000.0         0         1 2017-12-01  
5598782            Rice     64000.0         0         1 2017-12-01  
5735850            Rice     59990.0      3000         1 2017-12-01  
5678748            Rice     59500.0      3000         1 2017-12-01  
5702411            Rice     59500.0      3000         1 2017-12-01  

[12000 rows x 9 columns]

**NOTE:** Method `drop()` tidak mengubah objek dataframenya. Apabila ingin mengubah objek semulanya:
- Melakukan assignment kembali dengan nama objek yang sama: `rice = rice.drop(...)`, atau
- Menambahkan parameter inplace: `rice.drop(..., inplace=True)` 

In [ ]:
#your code here



### Slicing: **`[]` operator**

Digunakan untuk melakukan subsetting dengan cara mengiris (slicing) index pada dataframe. Formula penulisannya adalah `rice[start:end]` dengan mengikuti aturan indexing pada python (dimulai dari 0) dimana `start` inclusive dan `end` exclusive.

In [ ]:
#your code here



🧠 **Knowledge Check: Slicing**

Dengan memperhatikan `end` exclusive pada metode indexing, tampilkan baris ke 8 sampai ke 12 pada data rice. Pilih jawaban yang tepat dibawah ini!

- [ ] `rice[7:12]`
- [ ] `rice[8:12]`
- [ ] `rice[7:13]`
- [ ] `rice[8:13]`

#### Special Challenge

1. Lakukan indexing/subsetting untuk mendapatkan 5 data terakhir (kecuali dengan .tail())

In [ ]:
# code here


### `.loc` dan `.iloc`

Dengan menggunakan `.loc` dan `iloc` kita dapat melakukan pengirisan pada index **baris dan kolom**. 

Perbedaan yang mendasar dari kedua operator ini adalah:
- `.iloc` merujuk pada lokasi **index** baris atau kolomnya sehingga harus **integer**, sedangkan
- `.loc` merujuk pada lokasi **nama** baris atau kolomnya

#### .iloc

> Syntax: `df.iloc[baris, kolom]` 

Contoh: Kita ingin memilih baris pertama sampai kelima pada dataframe `rice`

In [ ]:
#your code here



#### Extra Challenge
Lakukan subsetting untuk menampilkan **2 baris terbawah** dan **4 kolom pertama**.

Gunakan beberapa metode subsetting dibawah ini :
1. Menggunakan `.iloc[ , ]`
2. Menggunakan method `tail(2)` untuk mendapatkan 2 baris terbawah, kemudian lakukan chaining dengan `.iloc`
3. Gunakan `.iloc` kemudian tambahkan `(rice.shape[0]-2):` didalamnya, untuk mendapatkan 2 baris terakhir.

In [ ]:
#your code here



#### .loc

> Syntax: `df.loc[baris, kolom]` 

Menggunakan `.loc`, kita bisa mengambil baris dan kolom berdasarkan namanya. Contoh kita ingin mengambil baris dengan `receipt_id` 9643416 dan 9706953, serta mengambil semua kolom. Gunakan DataFrame `rice_index`!

In [ ]:
#your code here
rice_index = pd.read_csv("data_input/rice.csv", index_col='receipt_id').drop(columns='Unnamed: 0')


In [ ]:
#your code here



#### Dive Deeper:
Baca data `companies.csv`dan gunakan `index_col=1`, kemudian tampilkan beberapa baris menggunakan method `head()` atau `tail()`

Dengan menggunakan metode indexing, jawab pertanyaan dibawah ini!

1. Tampilkan lokasi kantor *Li and Partners*?
2. Tampilkan nilai `Returns` yang dihasilkan oleh PT. Algoritma Data Indonesia?
3. Tampilkan nilai `Forecasted Growth` untuk Palembang Konsultasi?

In [ ]:
clients = pd.read_csv("data_input/companies.csv", index_col=1)
clients.head()

In [ ]:
#your code here



In [ ]:
#your code here



In [ ]:
#your code here



### Conditional Subsetting

Selain menggunakan `.loc` dan `.iloc`, kita dapat melakukan subsetting berdasarkan kondisi tertentu. Misal pada dataframe `rice`, kita ingin mengambil beberapa data dengan kondisi sebagai berikut:

- Transaksi yang terjadi di supermarket: `.format == 'supermarket'`
- Transaksi dengan produk yang harganya lebih dari sama dengan 200000: `.unit_price >= 200000`
- Transaksi dimana kuantitas tidak sama dengan 0: `.quantity != 0`

Syntax penulisan untuk conditional subsetting adalah:

**`df[df['column_name'] <comparison_operator> <value>]`**

atau

**`df[df.column_name <comparison_operator> <value>]`**

Contoh comparison_operator adalah seperti `==`, `!=`, `>`, `>=`, `<`, `<=`.

**Contoh 1** : Menampilkan data yang memiliki format Supermarket!

In [ ]:
#your code here



**Contoh 2** : Menampilkan data yang memiliki `unit_price` lebih besar sama dengan 200.000

In [ ]:
#your code here



❓ **Exercise** <br>
1. Tampilkan transaksi yang harganya (`unit_price`) kurang dari 100.000
2. Tampilkan transaksi selain di supermarket (kolom `format`)!

In [ ]:
#your code here



#### Multiple Condition

Kita juga dapat menggunakan operator `&` (AND) dan `|` (OR) untuk melakukan subsetting lebih dari 1 kondisi. Misalnya kita ingin melihat data penjualan dari seorang pegawai bernama Moana yang jumlahnya lebih dari 5000, maka kita dapat menggunakan syntax:
```
sales[(sales.salesperson == 'Moana') & (sales.amount > 5000)]
```

Untuk subsetting dengan kondisi lebih dari 1, setiap kondisi diletakkan **di dalam tanda kurung `()`** atau bisa ditulis dengan syntax berikut:

```
df[(kondisi pertama) operator (kondisi kedua) operator (kondisi ketiga) dan seterusnya...]
```

**Poin:**
- Operator AND: harus semua kondisi terpenuhi dalam satu baris agar muncul
- Operator OR: salah satu kondisi saja sudah terpenuhi maka baris tsb muncul
___
- cond1 = kondisi 1
- cond2 = kondisi 2
```
rice[(cond1)&(cond2)]
rice[(cond1)|(cond2)]
```

**Contoh 3**: Mengambil semua transaksi di `supermarket` yang `unit_price`nya di atas 200,000

In [ ]:
#your code here



#### Dive Deeper

Dengan menggunakan metode conditional subsetting, jawab pertanyaan dibawah ini:

1. Apabila Anda masih ingat, pada `.describe()` kita menemui bahwa terdapat transaksi janggal dengan **`discount` yang bernilai negatif**. Ada berapa transaksi yang janggal tersebut?
2. Perusahaan mendefinisikan pembelian grosir sebagai transaksi dengan `quantity` **minimal 12 units (1 Lusin)**. Berapa transaksi pada **hypermarket** yang dikatakan sebagai pembelian grosir?

*Hint: Gunakan `.shape` untuk mengetahui jumlah baris/kolom*

In [ ]:
#your code here



**Bonus Challenge**

Berapa banyak transaksi pada dataframe `rice` yang terjadi pada bulan Juli tahun 2018 baik di `minimarket` maupun `supermarket`?

In [ ]:
#your code here



# Referencing and Copying

Pada python, tanda sama dengan (`=`) bisa digunakan untuk me-**refer** sebuah Dataframe. Sekarang mari kita lihat dataframe `rice_data` berikut:

## Reference

Membuat dataframe baru dengan nama `rice_second` yang menyimpan semua transaksi seperti `rice_data`. Kemudian misalnya kita ingin update kolom `discount` agar menjadi 15 pada semua barisnya.

In [ ]:
rice_data = pd.read_csv("data_input/rice.csv", index_col=1)
rice_data = rice_data.drop('Unnamed: 0', axis=1)
rice_data.head(3)

In [ ]:
#your code here



## Copying

Membuat dataframe baru dengan nama `rice_third` yang menyimpan semua transaksi seperti `rice_data`. Kemudian misalnya kita ingin update kolom `discount` agar menjadi 15 pada semua barisnya.

In [ ]:
#your code here

